File: 3725.count-ways-to-choose-coprime-integers-from-rows.cpp
#
# @lc app=leetcode id=3725 lang=cpp
#
# [3725] Count Ways to Choose Coprime Integers from Rows
#
# @lc code=start
class Solution {
public:
    int countCoprime(vector<vector<int>>& mat) {
        const int MOD = 1e9 + 7;
        int m = mat.size();
        int n = mat[0].size();
        unordered_map<int, long long> gcd_count;
        // Initialize with first row values as potential gcds
        for (int num : mat[0]) {
            gcd_count[num]++;
        }
        // Process each subsequent row
        for (int i = 1; i < m; ++i) {
            unordered_map<int, long long> new_gcd_count;
            for (int num : mat[i]) {
                for (auto& [gcd_val, count] : gcd_count) {
                    int new_gcd = gcd(gcd_val, num);
                    new_gcd_count[new_gcd] = (new_gcd_count[new_gcd] + count) % MOD;
                }
                new_gcd_count[num] = (new_gcd_count[num] + 1) % MOD; // Count this number alone too.
            }
            gcd_count.swap(new_gcd_count); // Update with current row's results.
        }
        return gcd_count[1]; // Return number of ways to achieve gcd of 1.
    }
}; # @lc code=end

File: 3725.count-ways-to-choose-coprime-integers-from-rows.java
#
# @lc app=leetcode id=3725 lang=java
#
# [3725] Count Ways to Choose Coprime Integers from Rows
#

# @lc code=start
import java.util.*;
class Solution {
    public int countCoprime(int[][] mat) {
        int MOD = 1_000_000_007;
        int m = mat.length, n = mat[0].length;
        Map<Integer, Integer> dp = new HashMap<>();
        // Initialize DP with first row
        for (int val : mat[0]) {
            dp.put(val, dp.getOrDefault(val, 0) + 1);
        }
        // Process each row
        for (int i = 1; i < m; ++i) {
            Map<Integer, Integer> next = new HashMap<>();
            for (int val : mat[i]) {
                for (Map.Entry<Integer, Integer> entry : dp.entrySet()) {
                    int newGcd = gcd(val, entry.getKey());
                    next.put(newGcd, (int)(((long)next.getOrDefault(newGcd, 0) + entry.getValue()) % MOD));
                }
                // Also start new sequences from this row
                next.put(val, (int)(((long)next.getOrDefault(val, 0) + 1) % MOD));
            }
            dp = next;
        }
        return dp.getOrDefault(1, 0);
    }
    private int gcd(int a, int b) {
        while (b != 0) { int t = b; b = a % b; a = t; }
        return a;
    }
}
# @lc code=end